### PR TITLE
/home/guest/Desktop/corn_persam/per_segment_anything/predictor.py

### PR DESCRIPTION
**Pull request by JisuHann**
Three things are dealt here:
1. You can see the few-shot dataset for our setting [here](url). Please download to check in your setting
2. You can see the single shot version [here]( https://github.com/JisuHann/corn_persam/blob/few-shot/Personalize_SAM_with_one_shot_using_Hugging_Face_(SAM_f).ipynb)
3. I made a few-shot version [here](https://github.com/JisuHann/corn_persam/blob/few-shot/persam_f_fewshot.py) with the 참고 of [this single-shot version code](https://github.com/JisuHann/corn_persam/blob/few-shot/persam_f.py). Recommend you to compare both when reviewing code.

### CODE REVIEW

**Comment by yycho0108** [comment]
> Three things are dealt here:
> 
>     1. You can see the few-shot dataset for our setting [here](url). Please download to check in your setting
> 

this dataset url is clearly broken, pls fix

File: persam_f_fewshot.py [hunk 1]
```diff
@@ -23,12 +24,13 @@ def get_arguments():
     parser.add_argument('--data', type=str, default='./data')
     parser.add_argument('--outdir', type=str, default='persam_f')
     parser.add_argument('--ckpt', type=str, default='./sam_vit_h_4b8939.pth')
-    parser.add_argument('--sam_type', type=str, default='vit_h')
+    parser.add_argument('--sam_type', type=str, default='vit_t')
```
Comment (yycho0108) [inline persam_f_fewshot.py:27]: you're using default=vit_t but skill loading ckpt=sam_vit_h_4b8939
how does that make sense?

File: persam_f_fewshot.py [hunk 2]
```diff
@@ -39,86 +41,124 @@ def main():
     args = get_arguments()
     print("Args:", args)
 
-    images_path = args.data + '/Images/'
-    masks_path = args.data + '/Annotations/'
+    images_path = args.data
+    masks_path = args.data
```
Comment (yycho0108) [inline persam_f_fewshot.py:45]: What was the motivation behind combining these paths? Isn't it just more confusing?

File: persam_f_fewshot.py [hunk 3]
```diff
@@ -39,86 +41,124 @@ def main():
...
 
+    persam_f(args, 'example2', images_path, masks_path, output_path)
 
 def persam_f(args, obj_name, images_path, masks_path, output_path):
+
+    print("======> Load SAM" )
+    sam_type, sam_ckpt = 'vit_t', '/home/guest/Desktop/MobileSAM/weights/mobile_sam.pt'
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    sam = sam_model_registry[sam_type](checkpoint=sam_ckpt).to(device=device)
+    sam.eval()
```
Comment (yycho0108) [inline persam_f_fewshot.py:63]: you just made checkpoint loading unnecessarily worse by hardcoding your path instead of using the config arg as in the original code...

File: persam_f_fewshot.py [hunk 4]
```diff
@@ -39,86 +41,124 @@ def main():
...
+    gt_masks = []
+    ref_imgs_path = sorted(os.listdir(ref_image_path))
+    ref_masks_path = sorted(os.listdir(ref_mask_path))
     
+    for ref_img, ref_msk in zip(ref_imgs_path, ref_masks_path):    
+        ref_img_path = os.path.join(ref_image_path, ref_img)
+        ref_msk_path = os.path.join(ref_image_path, ref_msk)
+        ref_image = cv2.imread(ref_img_path)
+        ref_image = cv2.cvtColor(ref_image, cv2.COLOR_BGR2RGB)
+        ref_image = cv2.resize(ref_image, dsize=(1024,1024), interpolation=cv2.INTER_AREA)
```
Comment (yycho0108) [inline persam_f_fewshot.py:91]: original code doesn't seem to have `dsize()` resizing logic. Why is it needed now?

File: persam_f_fewshot.py [hunk 5]
```diff
@@ -39,86 +41,124 @@ def main():
...
+        transform_masks.append(transform_msk)
+    transform_imgs = torch.stack(transform_imgs)
+    transform_masks = torch.stack(transform_masks)
+
+    ref_mask = predictor.set_torch_image(transformed_image=transform_imgs, \
+                                         original_image_size = ref_img.shape[:2], \
+                                          transformed_mask=  transform_masks)
+    ref_feat = predictor.features.permute(0,2,3,1)
+    ref_mask = F.interpolate(ref_mask, size=ref_feat.shape[1: 3], mode="bilinear")
+    ref_mask = ref_mask[:, 0] # VALID. BUT WHY ZERO?
```
Comment (yycho0108) [inline persam_f_fewshot.py:129]: probably zero because it's taking the zeroth channel of an NCHW-formatted single-channel image
ref_mask.squeeze(dim=1) would also work, and return (B,H,W)

File: persam_f_fewshot.py [hunk 6]
```diff
@@ -39,86 +41,124 @@ def main():
...
     sim = F.interpolate(sim, scale_factor=4, mode="bilinear")
     sim = predictor.model.postprocess_masks(
                     sim,
                     input_size=predictor.input_size,
                     original_size=predictor.original_size).squeeze()
-
+    
     # Positive location prior
+    # Slightly modified to work in batched manner
     topk_xy, topk_label = point_selection(sim, topk=1)
```
Comment (yycho0108) [inline persam_f_fewshot.py:162]: if point_selection works in batched manner why does it still need unsqueeze(1)?

Comment (yycho0108) [reply persam_f_fewshot.py:162]: Also why are you using point inputs instead of mask inputs? This seems quite unnatural since you have masks and not points?

File: persam_f_fewshot.py [hunk 7]
```diff
@@ -260,17 +349,20 @@ def __init__(self):
...
+    topk_xy = mask_sim.flatten(1).topk(topk)[1]
+    topk_x = (topk_xy // h)
     topk_y = (topk_xy - topk_x * h)
-    topk_xy = torch.cat((topk_y, topk_x), dim=0).permute(1, 0)
-    topk_label = np.array([1] * topk)
-    topk_xy = topk_xy.cpu().numpy()
-    
+    print(f"\t\t\t{time.time()-start_time} seconds get h w")
+    start_time = time.time()
+    topk_xy = torch.cat((topk_y, topk_x), dim=1).to('cuda')
```
Comment (yycho0108) [inline persam_f_fewshot.py:361]: what happened to permute(1,0) in the above code? is that no longer necessary?

File: persam_f_fewshot.py [hunk 8]
```diff
@@ -260,17 +349,20 @@ def __init__(self):
...
+    print(f"\t\t\t{time.time()-start_time} seconds cat")
+    start_time = time.time()
+    topk_label = torch.tensor([1 for _ in range(b)], device='cuda')
```
Comment (yycho0108) [inline persam_f_fewshot.py:364]: This doesn't seem identical to np.array([1] * topk) (above code)

**Review by yycho0108 — COMMENTED** [review]
ok, the code is super obfuscated and I couldn't spot like a glaring error, but I suggest doing the following to help you debug:

* do _everything_ identical to persam_f.py, but just loop through the images manullay (with a for loop) instead of batchifying

Also, isn't the original code already few-shot? Otherwise what's the point of `images_path` being a directory in the original code rather than a filename?

**Comment by yycho0108** [comment]
Also if you're going to say "it doesn't work" I think it would be much more helpful if you can post the results and not just "it doesn't work"
